examples/sim_fibo_ad_cp.cpp

```diff
@@ -206,6 +207,38 @@ try
                                   *props, deck, gravity[2], state);
     }
 
+    std::shared_ptr<BlackoilState> distributed_state=Dune::stackobject_to_shared_ptr(state);
+    std::shared_ptr<Opm::BlackoilPropsAdFromDeck> distributed_props=new_props;
+    Dune::CpGrid distributed_grid = *grid;
+    // At this point all properties and state variables are correctly initialized
+    // If there are more than one processors involved, we now repartition the grid
+    // and initilialize new properties and states for it.
+    if( grid->comm().size()>=1 )
+    {
+        if(!param.getDefault("output_vtk", true))
+        {
+            OPM_THROW(std::logic_error, "We only support vtk output during parallel runs");
+        }
+        grid->loadBalance();
+        Dune::CpGrid global_grid      = *grid;
+        distributed_grid = *grid;
+        global_grid.switchToGlobalView();
+        distributed_grid.switchToDistributedView();
+        distributed_props = std::make_shared<BlackoilPropsAdFromDeck>(*new_props, grid->numCells());
+        distributed_state.reset(new BlackoilState);
+        distributed_state->init(grid->numCells(), grid->numFaces(), state.numPhases());
+        // init does not resize surfacevol. Do it manually.
+        distributed_state->surfacevol().resize(grid->numCells()*state.numPhases(),
+                                             std::numeric_limits<double>::max());
+        Opm::BlackoilStateDataHandle state_handle(global_grid, distributed_grid,
+                                                  state, *distributed_state);
+        Opm::BlackoilPropsDataHandle props_handle(global_grid, distributed_grid,
+                                                  static_cast<BlackoilPropsAdFromDeck&>(*new_props),
```

Comment:
You are right. Just saw that I changed the type in the ptr from the interface to the implementation.